File: Freedom/微信/基本类库/WXMessageManager.m
//  WXMessageManager.m
//  Freedom
// Created by Super
#import "WXMessageManager.h"
#import "WXChatViewController.h"
#import "WXUserHelper.h"
#import "WXModes.h"
@implementation WXImageMessage
- (CGSize)imageSize{
    CGFloat width = [[self.content objectForKey:@"w"] doubleValue];
    CGFloat height = [[self.content objectForKey:@"h"] doubleValue];
    return CGSizeMake(width, height);
}
- (void)setImageSize:(CGSize)imageSize{
    [self.content setObject:[NSNumber numberWithDouble:imageSize.width] forKey:@"w"];
    [self.content setObject:[NSNumber numberWithDouble:imageSize.height] forKey:@"h"];
}
#pragma mark -
- (WXMessageFrame *)messageFrame{
    if (kMessageFrame == nil) {
        kMessageFrame = [[WXMessageFrame alloc] init];
        kMessageFrame.height = 20 + (self.showTime ? 30 : 0) + (self.showName ? 15 : 0);
        CGSize imageSize = self.imageSize;
        if (CGSizeEqualToSize(imageSize, CGSizeZero)) {
            kMessageFrame.contentSize = CGSizeMake(100, 100);
        }else if (imageSize.width > imageSize.height) {
            CGFloat height = APPW * 0.45 * imageSize.height / imageSize.width;
            height = height < APPW * 0.25 ? APPW * 0.25 : height;
            kMessageFrame.contentSize = CGSizeMake(APPW * 0.45, height);
        }else{
            CGFloat width = APPW * 0.45 * imageSize.width / imageSize.height;
            width = width < APPW * 0.25 ? APPW * 0.25 : width;
            kMessageFrame.contentSize = CGSizeMake(width, APPW * 0.45);
        }
        
        kMessageFrame.height += kMessageFrame.contentSize.height;
    }
    return kMessageFrame;
}
- (NSString *)conversationContent{
    return @"[图片]";
}
- (NSString *)messageCopy{
    return [self.content mj_JSONString];
}
@end
@implementation WXConversation
- (void)setConvType:(TLConversationType)convType{
    _convType = convType;
    switch (convType) {
        case TLConversationTypePersonal:
        case TLConversationTypeGroup:
            _clueType = TLClueTypePointWithNumber;
            break;
        case TLConversationTypePublic:
        case TLConversationTypeServerGroup:
            _clueType = TLClueTypePoint;
            break;
        default:
            break;
    }
}
- (BOOL)isRead{
    return self.unreadCount == 0;
}
- (void)updateUserInfo:(WXUser *)user{
    self.partnerName = user.showName;
    self.avatarPath = user.avatarPath;
    self.avatarURL = user.avatarURL;
}
- (void)updateGroupInfo:(WXGroup *)group{
    self.partnerName = group.groupName;
    self.avatarPath = group.groupAvatarPath;
}
@end
@implementation WXMessageFrame
@end
@implementation WXMessage
+ (WXMessage *)createMessageByType:(TLMessageType)type{
    NSString *className;
    if (type == TLMessageTypeText) {
        className = @"TLTextMessage";
    }else if (type == TLMessageTypeImage) {
        className = @"TLImageMessage";
    }else if (type == TLMessageTypeExpression) {
        className = @"TLExpressionMessage";
    }
    if (className) {
        return [[NSClassFromString(className) alloc] init];
    }
    return nil;
}
- (id)init{
    if (self = [super init]) {
        self.messageID = [NSString stringWithFormat:@"%lld", (long long)([[NSDate date] timeIntervalSince1970] * 10000)];
    }
    return self;
}
#pragma mark - Protocol
- (NSString *)conversationContent{
    return @"子类未定义";
}
- (NSString *)messageCopy{
    return @"子类未定义";
}
#pragma mark -
- (NSMutableDictionary *)content{
    if (_content == nil) {
        _content = [[NSMutableDictionary alloc] init];
    }
    return _content;
}
@end
static WXMessageManager *messageManager;
@implementation WXMessageManager
+ (WXMessageManager *)sharedInstance{
    static dispatch_once_t once;
    dispatch_once(&once, ^{
        messageManager = [[WXMessageManager alloc] init];
    });
    return messageManager;
}
- (void)sendMessage:(WXMessage *)message
           progress:(void (^)(WXMessage *, CGFloat))progress
            success:(void (^)(WXMessage *))success
            failure:(void (^)(WXMessage *))failure{
    BOOL ok = [self.messageStore addMessage:message];
    if (!ok) {
        DLog(@"存储Message到DB失败");
    }else{      // 存储到conversation
        ok = [self addConversationByMessage:message];
        if (!ok) {
            DLog(@"存储Conversation到DB失败");
        }
    }
}
#pragma mark - Getter -
- (WXDBMessageStore *)messageStore{
    if (_messageStore == nil) {
        _messageStore = [[WXDBMessageStore alloc] init];
    }
    return _messageStore;
}
- (WXDBConversationStore *)conversationStore{
    if (_conversationStore == nil) {
        _conversationStore = [[WXDBConversationStore alloc] init];
    }
    return _conversationStore;
}
- (NSString *)userID{
    return [WXUserHelper sharedHelper].user.userID;
}
- (BOOL)addConversationByMessage:(WXMessage *)message{
    NSString *partnerID = message.friendID;
    NSInteger type = 0;
    if (message.partnerType == TLPartnerTypeGroup) {
        partnerID = message.groupID;
        type = 1;
    }
    BOOL ok = [self.conversationStore addConversationByUid:message.userID fid:partnerID type:type date:message.date];
    
    return ok;
}
- (void)conversationRecord:(void (^)(NSArray *))complete{
    NSArray *data = [self.conversationStore conversationsByUid:self.userID];
    complete(data);
}
- (BOOL)deleteConversationByPartnerID:(NSString *)partnerID{
    BOOL ok = [self deleteMessagesByPartnerID:partnerID];
    if (ok) {
        ok = [self.conversationStore deleteConversationByUid:self.userID fid:partnerID];
    }
    return ok;
}
- (void)messageRecordForPartner:(NSString *)partnerID
                       fromDate:(NSDate *)date
                          count:(NSUInteger)count
                       complete:(void (^)(NSArray *, BOOL))complete{
    [self.messageStore messagesByUserID:self.userID partnerID:partnerID fromDate:date count:count complete:^(NSArray *data, BOOL hasMore) {
        complete(data, hasMore);
    }];
}
- (void)chatFilesForPartnerID:(NSString *)partnerID
                    completed:(void (^)(NSArray *))completed{
    NSArray *data = [self.messageStore chatFilesByUserID:self.userID partnerID:partnerID];
    completed(data);
}
- (void)chatImagesAndVideosForPartnerID:(NSString *)partnerID
                              completed:(void (^)(NSArray *))completed
{
    NSArray *data = [self.messageStore chatImagesAndVideosByUserID:self.userID partnerID:partnerID];
    completed(data);
}
- (BOOL)deleteMessageByMsgID:(NSString *)msgID{
    return [self.messageStore deleteMessageByMessageID:msgID];
}
- (BOOL)deleteMessagesByPartnerID:(NSString *)partnerID{
    BOOL ok = [self.messageStore deleteMessagesByUserID:self.userID partnerID:partnerID];
    if (ok) {
        [[WXChatViewController sharedChatVC] resetChatVC];
    }
    return ok;
}
- (BOOL)deleteAllMessages{
    BOOL ok = [self.messageStore deleteMessagesByUserID:self.userID];
    if (ok) {
        [[WXChatViewController sharedChatVC] resetChatVC];
        ok = [self.conversationStore deleteConversationsByUid:self.userID];
    }
    return ok;
}
- (void)requestClientInitInfoSuccess:(void (^)(id)) clientInitInfo
                             failure:(void (^)(NSString *))error{
    NSString *HOST_URL = @"http://127.0.0.1:8000/";            // 本地测试服务器
//    HOST_URL = @"http://121.42.29.15:8000/";        // 远程线上服务器
    NSString *urlString = [HOST_URL stringByAppendingString:@"client/getClientInitInfo/"];
    [[AFHTTPSessionManager manager] POST:urlString parameters:nil progress:nil success:^(NSURLSessionDataTask * _Nonnull task, id  _Nullable responseObject) {
        DLog(@"OK");
    } failure:^(NSURLSessionDataTask * _Nullable task, NSError * _Nonnull error) {
        DLog(@"OK");
    }];
}
- (void)userLoginWithUsername:(NSString *)username
                     password:(NSString *)password
                      success:(void (^)(id))userInfo
                      failure:(void (^)(NSString *))error{
}
- (NSMutableArray *)chatDetailDataByUserInfo:(WXUser *)userInfo{
    WXSettingItem *users = TLCreateSettingItem(@"users");
    users.type = TLSettingItemTypeOther;
    WXSettingGroup *group1 = TLCreateSettingGroup(nil, nil, @[users]);
    
    WXSettingItem *top = TLCreateSettingItem(@"置顶聊天");
    top.type = TLSettingItemTypeSwitch;
    WXSettingItem *screen = TLCreateSettingItem(@"消息免打扰");
    screen.type = TLSettingItemTypeSwitch;
    WXSettingGroup *group2 = TLCreateSettingGroup(nil, nil, (@[top, screen]));
    
    WXSettingItem *chatFile = TLCreateSettingItem(@"聊天文件");
    WXSettingGroup *group3 = TLCreateSettingGroup(nil, nil, @[chatFile]);
    
    WXSettingItem *chatBG = TLCreateSettingItem(@"设置当前聊天背景");
    WXSettingItem *chatHistory = TLCreateSettingItem(@"查找聊天内容");
    WXSettingGroup *group4 = TLCreateSettingGroup(nil, nil, (@[chatBG, chatHistory]));
    
    WXSettingItem *clear = TLCreateSettingItem(@"清空聊天记录");
    clear.showDisclosureIndicator = NO;
    WXSettingGroup *group5 = TLCreateSettingGroup(nil, nil, @[clear]);
    
    WXSettingItem *report = TLCreateSettingItem(@"举报");
    WXSettingGroup *group6 = TLCreateSettingGroup(nil, nil, @[report]);
    
    NSMutableArray *data = [[NSMutableArray alloc] init];
    [data addObjectsFromArray:@[group1, group2, group3, group4, group5, group6]];
    return data;
}
- (NSMutableArray *)chatDetailDataByGroupInfo:(WXGroup *)groupInfo{
    WXSettingItem *users = TLCreateSettingItem(@"users");
    users.type = TLSettingItemTypeOther;
    WXSettingItem *allUsers = TLCreateSettingItem(([NSString stringWithFormat:@"全部群成员(%ld)", (long)groupInfo.count]));
    WXSettingGroup *group1 = TLCreateSettingGroup(nil, nil, (@[users, allUsers]));
    
    WXSettingItem *groupName = TLCreateSettingItem(@"群聊名称");
    groupName.subTitle = groupInfo.groupName;
    WXSettingItem *groupQR = TLCreateSettingItem(@"群二维码");
    groupQR.rightImagePath = PQRCode;
    WXSettingItem *groupPost = TLCreateSettingItem(@"群公告");
    if (groupInfo.post.length > 0) {
        groupPost.subTitle = groupInfo.post;
    }else{
        groupPost.subTitle = @"未设置";
    }
    WXSettingGroup *group2 = TLCreateSettingGroup(nil, nil, (@[groupName, groupQR, groupPost]));
    
    WXSettingItem *screen = TLCreateSettingItem(@"消息免打扰");
    screen.type = TLSettingItemTypeSwitch;
    WXSettingItem *top = TLCreateSettingItem(@"置顶聊天");
    top.type = TLSettingItemTypeSwitch;
    WXSettingItem *save = TLCreateSettingItem(@"保存到通讯录");
    save.type = TLSettingItemTypeSwitch;
    WXSettingGroup *group3 = TLCreateSettingGroup(nil, nil, (@[screen, top, save]));
    
    WXSettingItem *myNikeName = TLCreateSettingItem(@"我在本群的昵称");
    myNikeName.subTitle = groupInfo.myNikeName;
    WXSettingItem *showOtherNikeName = TLCreateSettingItem(@"显示群成员昵称");
    showOtherNikeName.type = TLSettingItemTypeSwitch;
    WXSettingGroup *group4 = TLCreateSettingGroup(nil, nil, (@[myNikeName, showOtherNikeName]));
    
    WXSettingItem *chatFile = TLCreateSettingItem(@"聊天文件");
    WXSettingItem *chatHistory = TLCreateSettingItem(@"查找聊天内容");
    WXSettingItem *chatBG = TLCreateSettingItem(@"设置当前聊天背景");
    WXSettingItem *report = TLCreateSettingItem(@"举报");
    WXSettingGroup *group5 = TLCreateSettingGroup(nil, nil, (@[chatFile, chatHistory, chatBG, report]));
    
    WXSettingItem *clear = TLCreateSettingItem(@"清空聊天记录");
    clear.showDisclosureIndicator = NO;
    WXSettingGroup *group6 = TLCreateSettingGroup(nil, nil, @[clear]);
    
    NSMutableArray *data = [[NSMutableArray alloc] init];
    [data addObjectsFromArray:@[group1, group2, group3, group4, group5, group6]];
    return data;
}
@end

File: Freedom/微信/基本类库/WXMessageManager.swift
//
//  WXMessageManager.swift
//  Freedom
import Realm
import RealmSwift
import Foundation
//消息所有者类型
enum TLPartnerType : Int {
    case user // 用户
    case group // 群聊
}
//消息拥有者
enum TLMessageOwnerType : Int {
    case unknown // 未知的消息拥有者
    case system // 系统消息
    case own // 自己发送的消息
    case friend // 接收到的他人消息
}
//消息发送状态
enum TLMessageSendState : Int {
    case success // 消息发送成功
    case fail // 消息发送失败
}
//消息读取状态
enum TLMessageReadState : Int {
    case unRead // 消息未读
    case readed // 消息已读
}
protocol WXMessageProtocol: NSObjectProtocol {
    func messageCopy() -> String
    func conversationContent() -> String
}
protocol WXMessageManagerConvVCDelegate: NSObjectProtocol {
    func updateConversationData()
}
@objcMembers
class WXConversation: RLMObject {
    var remindType = TLMessageRemindType.normal//消息提醒类型
    var partnerID = ""//用户ID
    var partnerName = ""//用户名
    var avatarURL = ""//头像地址（网络）
    var avatarPath = ""//头像地址（本地）
    var date: Date = Date()//时间
    var content = ""//消息展示内容
    var clueType = TLClueType.none//提示红点类型
    var unreadCount: Int = 0//未读数
    var isRead: Bool {
        return unreadCount == 0
    }
    //会话类型（个人，讨论组，企业号）
    var convType = TLConversationType.personal {
        didSet {
            switch convType {
            case .personal, .group:
                clueType = .pointWithNumber
            case .public, .serverGroup:
                clueType = .point
            }
        }
    }
    override static func primaryKey() -> String {
        return "partnerID"
    }
    func updateUserInfo(_ user: WXUser) {
        partnerName = user.showName
        avatarPath = user.avatarPath
        avatarURL = user.avatarURL
    }
    func updateGroupInfo(_ group: WXGroup) {
        partnerName = group.groupName
        avatarPath = group.groupAvatarPath
    }
}
@objcMembers
class WXMessageFrame: RLMObject {
    var height: CGFloat = 0
    var width: CGFloat = 0
    var contentSize: CGSize {
        set{
            height = contentSize.height
            width = contentSize.width
        }
        get{
            return CGSize(width: width, height: height)
        }
    }

    override static func ignoredProperties() -> [String] {
        return ["contentSize"]
    }
}
@objcMembers
class WXMessageContent: RLMObject {
    var w: CGFloat = 0
    var h: CGFloat = 0
    var text = ""
    var url = ""
    var path = ""
    var groupID = ""
    var emojiID = ""
    override static func ignoredProperties() -> [String] {
        return ["w","h","text","url","path","groupID","emojiID"]
    }
}
@objcMembers
class WXMessage: RLMObject, WXMessageProtocol {
    var kMessageFrame: WXMessageFrame!
    var fromUser: WXChatUserProtocol?
    var messageFrame: WXMessageFrame!
    var messageID = ""
    var userID = ""
    var friendID = ""
    var groupID = ""
    var date = Date()
    var showTime = false
    var showName = false
    var partnerType = TLPartnerType.user
    var messageType = TLMessageType.text
    var ownerTyper = TLMessageOwnerType.own
    var readState = TLMessageReadState.unRead
    var sendState = TLMessageSendState.success
    var content: WXMessageContent!
    override static func primaryKey() -> String {
        return "messageID"
    }
    override static func ignoredProperties() -> [String] {
        return ["content"]
    }
    class func createMessage(by type: TLMessageType) -> WXMessage {
        if type == .text {
            return WXTextMessage()
        } else if type == .image {
            return WXImageMessage()
        } else if type == .expression {
            return WXExpressionMessage()
        }
        return WXMessage()
    }
    override init() {
        super.init()
        kMessageFrame = WXMessageFrame()
        messageFrame = WXMessageFrame()
        content = WXMessageContent()
        messageID = String(format: "%lld", Int64(Date().timeIntervalSince1970 * 10000))
    }
    func conversationContent() -> String {
        return "子类未定义"
    }
    func messageCopy() -> String {
        return "子类未定义"
    }
}
class WXTextMessage: WXMessage {
    let textLabel = UILabel()
    var attrText: NSAttributedString {return NSAttributedString(string: content.text)}
    override init() {
        super.init()
        kMessageFrame.height = 20 + (showTime ? 30 : 0) + (showName ? 15 : 0)
        if messageType == .text {
            kMessageFrame.height += 20
            textLabel.attributedText = attrText
            kMessageFrame.contentSize = textLabel.sizeThatFits(CGSize(width: APPW * 0.58, height: CGFloat(MAXFLOAT)))
        }
        kMessageFrame.height += kMessageFrame.contentSize.height

        var size = CGFloat(UserDefaults.standard.double(forKey: "CHAT_FONT_SIZE"))
        if size == 0 {
            size = 16.0
        }
        textLabel.font = UIFont.systemFont(ofSize: size)
        textLabel.numberOfLines = 0
    }
    override func conversationContent() -> String {
        return content.text
    }
    override func messageCopy() -> String {
        return content.text
    }
}

class WXImageMessage: WXMessage {
    var imageSize: CGSize {
        set (newValue){
            content.w = imageSize.width
            content.h = imageSize.height
        }
        get {
            let width = content.w
            let height = content.h
            return CGSize(width: width, height: height)
        }
    }
    override init() {
        super.init()
        kMessageFrame.height = 20 + (showTime ? 30 : 0) + (showName ? 15 : 0)
        let imageSize: CGSize = self.imageSize
        if imageSize.equalTo(CGSize.zero) {
            kMessageFrame.contentSize = CGSize(width: 100, height: 100)
        } else if imageSize.width > imageSize.height {
            var height: CGFloat = APPW * 0.45 * imageSize.height / imageSize.width
            height = height < APPW * 0.25 ? APPW * 0.25 : height
            kMessageFrame.contentSize = CGSize(width: APPW * 0.45, height: height)
        } else {
            var width: CGFloat = APPW * 0.45 * imageSize.width / imageSize.height
            width = width < APPW * 0.25 ? APPW * 0.25 : width
            kMessageFrame.contentSize = CGSize(width: width, height: APPW * 0.45)
        }
        kMessageFrame.height += kMessageFrame.contentSize.height

    }
    override func conversationContent() -> String {
        return "[图片]"
    }
    override func messageCopy() -> String {
        return content.text
    }
}

class WXExpressionMessage: WXMessage {
    var emoji: TLEmoji! {
        didSet {
            content["groupID"] = emoji.groupID
            content["emojiID"] = emoji.emojiID
            let imageSize: CGSize = UIImage(named: self.path)?.size ?? CGSize.zero
            content["w"] = "\(imageSize.width)"
            content["h"] = "\(imageSize.height)"
        }
    }
    var path: String {
        set {
//            emoji.emojiPath = path
            print(path)
        }
        get {
            return emoji.emojiPath
        }
    }
    var url:String {
        return "http://123.57.155.230:8080/ibiaoqing/admin/expre/download.dopId=\(emoji.emojiID)"
    }
    var emojiSize: CGSize {
        set {
            content.w = newValue.width
            content.h = newValue.height
        }
        get {
            return CGSize(width: content.w, height: content.h)
        }
    }
    override init() {
        super.init()
        emoji = TLEmoji()
        emoji.groupID = content.groupID
        emoji.emojiID = content.emojiID
        kMessageFrame.height = 20 + (showTime ? 30 : 0) + (showName ? 15 : 0)
        kMessageFrame.height += 5
        let emojiSize: CGSize = CGSize.zero
        if emojiSize.equalTo(CGSize.zero) {
            kMessageFrame.contentSize = CGSize(width: 80, height: 80)
        } else if emojiSize.width > emojiSize.height {
            var height: CGFloat = APPW * 0.35 * emojiSize.height / emojiSize.width
            height = height < APPW * 0.2 ? APPW * 0.2 : height
            kMessageFrame.contentSize = CGSize(width: APPW * 0.35, height: height)
        } else {
            var width: CGFloat = APPW * 0.35 * emojiSize.width / emojiSize.height
            width = width < APPW * 0.2 ? APPW * 0.2 : width
            kMessageFrame.contentSize = CGSize(width: width, height: APPW * 0.35)
        }
        kMessageFrame.height += kMessageFrame.contentSize.height
    }
    override func conversationContent() -> String {
        return "[表情]"
    }
    override func messageCopy() -> String {
        return content.text
    }
}

class WXMessageManager: NSObject {
    static let shared = WXMessageManager()
    var messageDelegate: Any?
    weak var conversationDelegate: WXMessageManagerConvVCDelegate?
    private var userID: String {
        return WXUserHelper.shared.user.userID
    }
    func send(_ message: WXMessage, progress: @escaping (WXMessage, CGFloat) -> Void, success: @escaping (WXMessage) -> Void, failure: @escaping (WXMessage) -> Void) {
        WXDBStore.shared.add(message)
        addConversation(by: message)
    }
    func addConversation(by message: WXMessage) {
        var partnerID = message.friendID
        var type: Int = 0
        if message.partnerType == .group {
            partnerID = message.groupID
            type = 1
        }
        WXDBStore.shared.addConversation(byUid: message.userID, fid: partnerID, type: type, date: message.date)
    }
    func conversationRecord(_ complete: @escaping ([WXConversation]) -> Void) {
        let data = WXDBStore.shared.conversations(byUid: userID)
        complete(data)
    }

    func deleteConversation(byPartnerID partnerID: String) {
        deleteMessages(byPartnerID: partnerID)
        WXDBStore.shared.deleteConversation(byUid: userID, fid: partnerID)
    }

    func messageRecord(forPartner partnerID: String, from date: Date, count: Int, complete: @escaping ([WXMessage], Bool) -> Void) {
        WXDBStore.shared.messages(byUserID: userID, partnerID: partnerID, from: date, count: count, complete: { data, hasMore in
            complete(data, hasMore)
        })
    }
    func chatFiles(forPartnerID partnerID: String, completed: @escaping ([[WXMessage]]) -> Void) {
        let data = WXDBStore.shared.chatFiles(byUserID: userID, partnerID: partnerID)
        completed(data)
    }

    func chatImagesAndVideos(forPartnerID partnerID: String, completed: @escaping ([WXImageMessage]) -> Void) {
        let data = WXDBStore.shared.chatImagesAndVideos(byUserID: userID, partnerID: partnerID)
        completed(data)
    }

    func deleteMessage(byMsgID msgID: String) {
        WXDBStore.shared.deleteMessage(byMessageID: msgID)
    }

    func deleteMessages(byPartnerID partnerID: String) {
        WXDBStore.shared.deleteMessages(byUserID: userID, partnerID: partnerID)
        WXChatViewController.shared.resetChatVC()
    }
    func deleteAllMessages() {
        WXDBStore.shared.deleteMessages(byUserID: userID)
        WXChatViewController.shared.resetChatVC()
        WXDBStore.shared.deleteConversations(byUid: userID)
    }

    func requestClientInitInfoSuccess(_ clientInitInfo: @escaping (Any) -> Void, failure error: @escaping (String) -> Void) {
        let HOST_URL = "http://127.0.0.1:8000/" // 本地测试服务器
        //    HOST_URL = @"http://121.42.29.15:8000/";        // 远程线上服务器
        let urlString = HOST_URL + ("client/getClientInitInfo/")
        AFHTTPSessionManager().post(urlString, parameters: nil, progress: nil, success: { task, responseObject in
            Dlog("OK")
        }, failure: { task, error in
            Dlog("OK")
        })
    }
    func chatDetailData(byUserInfo userInfo: WXUser) -> [WXSettingGroup] {
        let users = WXSettingItem.createItem(withTitle: "users")
        users.type = .other
        let group1: WXSettingGroup = TLCreateSettingGroup(nil, nil, [users])
        let top = WXSettingItem.createItem(withTitle: ("置顶聊天"))
        top.type = .switchBtn
        let screen = WXSettingItem.createItem(withTitle: ("消息免打扰"))
        screen.type = .switchBtn
        let group2: WXSettingGroup = TLCreateSettingGroup(nil, nil, ([top, screen]))
        let chatFile = WXSettingItem.createItem(withTitle: ("聊天文件"))
        let group3: WXSettingGroup = TLCreateSettingGroup(nil, nil, [chatFile])
        let chatBG = WXSettingItem.createItem(withTitle: ("设置当前聊天背景"))
        let chatHistory = WXSettingItem.createItem(withTitle: ("查找聊天内容"))
        let group4: WXSettingGroup = TLCreateSettingGroup(nil, nil, ([chatBG, chatHistory]))
        let clear = WXSettingItem.createItem(withTitle: ("清空聊天记录"))
        clear.showDisclosureIndicator = false
        let group5: WXSettingGroup = TLCreateSettingGroup(nil, nil, [clear])
        let report = WXSettingItem.createItem(withTitle: ("举报"))
        let group6: WXSettingGroup = TLCreateSettingGroup(nil, nil, [report])
        return [group1, group2, group3, group4, group5, group6]
    }

    func chatDetailData(byGroupInfo groupInfo: WXGroup) -> [WXSettingGroup] {
        let users = WXSettingItem.createItem(withTitle: ("users"))
        users.type = .other
        let allUsers = WXSettingItem.createItem(withTitle: (String(format: "全部群成员(%ld)", Int(groupInfo.count))))
        let group1: WXSettingGroup = TLCreateSettingGroup(nil, nil, ([users, allUsers]))
        let groupName = WXSettingItem.createItem(withTitle: ("群聊名称"))
        groupName.subTitle = groupInfo.groupName
        let groupQR = WXSettingItem.createItem(withTitle: ("群二维码"))
        groupQR.rightImagePath = PQRCode
        let groupPost = WXSettingItem.createItem(withTitle: ("群公告"))
        if groupInfo.post.count > 0 {
            groupPost.subTitle = groupInfo.post
        } else {
            groupPost.subTitle = "未设置"
        }
        let group2: WXSettingGroup = TLCreateSettingGroup(nil, nil, ([groupName, groupQR, groupPost]))
        let screen = WXSettingItem.createItem(withTitle: ("消息免打扰"))
        screen.type = .switchBtn
        let top = WXSettingItem.createItem(withTitle: ("置顶聊天"))
        top.type = .switchBtn
        let save = WXSettingItem.createItem(withTitle: ("保存到通讯录"))
        save.type = .switchBtn
        let group3: WXSettingGroup = TLCreateSettingGroup(nil, nil, ([screen, top, save]))
        let myNikeName = WXSettingItem.createItem(withTitle: ("我在本群的昵称"))
        myNikeName.subTitle = groupInfo.myNikeName
        let showOtherNikeName = WXSettingItem.createItem(withTitle: ("显示群成员昵称"))
        showOtherNikeName.type = .switchBtn
        let group4: WXSettingGroup = TLCreateSettingGroup(nil, nil, ([myNikeName, showOtherNikeName]))
        let chatFile = WXSettingItem.createItem(withTitle: ("聊天文件"))
        let chatHistory = WXSettingItem.createItem(withTitle: ("查找聊天内容"))
        let chatBG = WXSettingItem.createItem(withTitle: ("设置当前聊天背景"))
        let report = WXSettingItem.createItem(withTitle: ("举报"))
        let group5: WXSettingGroup = TLCreateSettingGroup(nil, nil, ([chatFile, chatHistory, chatBG, report]))
        let clear = WXSettingItem.createItem(withTitle: ("清空聊天记录"))
        clear.showDisclosureIndicator = false
        let group6: WXSettingGroup = TLCreateSettingGroup(nil, nil, [clear])
        return [group1, group2, group3, group4, group5, group6]
    }
}
